### DISCUSSION TITLE
How to use a condition in Squib with If or Case?

### DISCUSSION THREAD

**Discussion by LouisDeconinck** [andymeneely/squib#354] [Q&A] [answered]
The following code always returns 'No sun' even if the SunUp variable contains 1 or 2. I noticed that this is because data['SunUp'] is always an array. How can I change certain elements on the cards if a certain condition is met? Using ranges (https://squib.readthedocs.io/en/v0.18.0/arrays.html) seems a lot less flexible. 

```
 case data['SunUp']
  when "1"
    png file: 'images/Sun-256.png', x: 300, width: 100, height: :scale
  when "2"
    png file: 'images/Sun-256.png', x: 300, width: 100, height: :scale
    png file: 'images/Sun-256.png', x: 420, width: 100, height: :scale
  else
    text str: 'No sun', x: 300
  end
  
  text str: data['SunUp']
```

In the example (https://squib.readthedocs.io/en/v0.18.0/arrays.html#examples), this code is used to identify the Thief and  Grifter: `text range: 0..1, str: 'Thief and Grifter only!!', x: 25, y:200` Wouldn't it be a lot more logical to check if the name of the card matches Thief or Grifter? That way, if the cards change position in your data, the code still works.

Is there any possibility of using case or if when using Squib? If not this would seem like a very useful function.

EDIT: For now I solved the problem by generating new image files for each combination of Sun and other powers my cards can have. This is very inefficient because I have to create images for each combination of 3 elements with 5 varieties. Then I dynamically generate the file name in my spreadsheet. So, still looking for a more elegant solution, even if this did work.

**Comment by andymeneely** [comment]
You want to apply the conditional on each element of the array, so it needs to be inside of an iterator. I tend to use `map` a lot more often than ranges. So something like this:

```ruby
png file: data['SunUp'].map {|s| ... } 
```

There's an example of doing a different number of images given a number on that array page, starting at line 52.

I've seen some folks use the `text` and icon embedding. So something like this:
```ruby
suns = data['SunUp'].map { |s| s.to_i * ":sun:" }
text str: suns, key: ':sun:' do |embed|
  embed.png ...
end
```

**Reply by LouisDeconinck** [reply]
This seems like a solution I'll test it out. I also went through one of your projects and it seemed like you used `ENV` to accomplish the same, what does that do?

**Reply by LouisDeconinck** [reply]
I'm still struggling with this
> ```ruby
> png file: data['SunUp'].map {|s| ... } 
> ```

How would this work if I nil was a possibility? I tried using:
```png file: data['Effect1'].map { |t| "images/#{t.downcase}1.png"}, x: 100, y: 800, width: :scale, height: 150```

This gives me the error: "downcase does not work for nil". If I add ```.compact``` that results in effects shifting from one card to another card. I just want the script to ignore and not put an image when the value is nil. How to accomplish this?

Is there really no way of adding an if statement to check if certain conditions are met for each iteration over the array?

**Reply by andymeneely** [reply]
Oops put my response in a separate answer. See below

**Comment by andymeneely** [comment, marked as answer]
There's always a way!

Ruby's `nil` has the `to_s` method and it returns `''`, so I would do `"images/#{t.to_s.downcase}1.png"`

Ruby also has the `&.` operator as well. So `t&.downcase` would return `nil` if `t` was nil. https://ruby-doc.org/core-2.6/doc/syntax/calling_methods_rdoc.html

Either of those options along with the new `placeholder` option to png would be good.

You can do if-statements inside of map. So things like 

```
data.effect1.map do |e|
  if something
    foo
  else 
    bar
end
```

Uglier option is the ternary operator... so like `data.effect1.map { |e| e.something ? foo : bar }`

**Reply by LouisDeconinck** [reply]
Using the `to_s` with placeholder method works. As a placeholder, I put a random `"a"`. While it does work, it does not look like clean programming, on top of that, it also throws a bunch of warnings. Any possibility of preventing this?

The if method works as well, but also gives me warnings when it tries to display an image using `nil`. However, I do not understand that. This is my code:
```  
data['Effect1'].map do |e|
 if data['Effect1'] != nil
  png file: data['Effect1'].map { |t| "images/#{t.to_s.downcase}1.png"}, x: 110, y: 800, width: :scale , height: 150
 end
end
```

How come, that if Effect1 contains `nil` it still tries to generate an image? I just want it to skip over it.

Another strange thing I noticed, is that I get this warning about nil, 3 times (3 cards), even though only one card has this as nil.

**Reply by LouisDeconinck** [reply]
@andymeneely If you would find some time to help me out, that would be great!

**Reply by LouisDeconinck** [reply]
I did not fully understand how mapping worked, in the example above I did a check on the entire array instead of one element of the array. I should have used `e` instead of `data['Effect1]`. 

The following code did work for me:
```
png file: data['Effect1'].map { |t| t.nil? ? nil : "images/#{t.to_s.downcase}.png"}, x: 105, y: 800, width: :scale , height: 150
```
